### DISCUSSION TITLE
Make `typer.Context` generic for easier typing `Context.obj`

### DISCUSSION THREAD

**Discussion by Eggiverse** [fastapi/typer#1064] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import Typer

app = typer.Typer()

@dataclass
class MyObj:
    prop: int
# Currently
@app.command()
def my_command(ctx: typer.Context):
    obj: MyObj = ctx.obj # I have to specify type of object here
    print(obj.prop)

# If `typer.Context` is generic
@app.command()
def my_command(ctx: typer.Context[MyObj]):
    obj = ctx.obj # type is automatically inferred
    print(obj.prop)

# Possible solution. Not tested yet.
# Modify `typer.Context` as
class Context[T](click.Context):
    obj: T
```

### Description

I used to use click and frequently use `click.pass_obj` since I do not use other features in `Context`. I usually type my object in the command definition like this:
```python
class MyObj:
    ...

@click.command()
@click.pass_obj
def my_command(obj: MyObj, ...):
    ... # My type checker is aware that `obj` is of type `MyObj`
```

In Typer, I would have to do this:
```python
@app.command()
def my_command(ctx: typer.Context, ...):
    obj: MyObj = ctx.obj # I have to specify type of object here
    print(obj.prop)
```
I believe it would be better if I can still specify object type in type annotation. Making `typing.Conext` generic might be a solution.
```python
@app.command()
def my_command(ctx: typer.Context[MyObj], ...):
    obj = ctx.obj # type is automatically inferred
    print(obj.prop)
```

### Operating System

Windows

### Typer Version

0.13.1

### Python Version

3.12

**Comment by AtticusZeller** [comment]
same problem
